utils/vfs/vfs.cpp

```diff
@@ -0,0 +1,123 @@
+#include <algorithm>
+#include <fuzzuf/utils/filesystem.hpp>
+#include <fuzzuf/utils/vfs/vfs.hpp>
+#include <fuzzuf/exceptions.hpp>
+namespace fuzzuf::utils::vfs {
+
+VFS::VFS(
+  std::vector< fs::path > &&allowed_path_
+) {
+  if( !allowed_path_.empty() ) {
+    std::vector< fs::path > duplicated_allowed_path;
+    duplicated_allowed_path.reserve( allowed_path_.size() );
+    if( !allowed_path_[ 0 ].has_root_directory() )
+      throw exceptions::invalid_argument( "First allowed path must contain root directory", __FILE__, __LINE__ );
+    {
+      auto can = allowed_path_[ 0 ].lexically_normal();
+      current_workdir = can;
+    } 
+    std::transform(
+      allowed_path_.begin(),
+      allowed_path_.end(),
+      std::back_inserter( duplicated_allowed_path ),
+      [this]( const auto &v ) {
+        auto abs = Absolute( v );
+        auto can = abs.lexically_normal();
+        return can;
+      }
+    );
+    std::sort(
+      duplicated_allowed_path.begin(),
+      duplicated_allowed_path.end()
+    );
+    allowed_path.reset( new std::vector< fs::path >{} );
+    allowed_path->push_back( std::move( duplicated_allowed_path[ 0 ] ) );
+    current_workdir = (*allowed_path)[ 0 ];
+    fs::path prev = duplicated_allowed_path[ 0 ];
+    auto prev_len = std::distance( prev.begin(), prev.end() );
+    for(
+      auto cur = std::next( duplicated_allowed_path.begin() );
+      cur != duplicated_allowed_path.end();
+      ++cur
+    ) {
+      auto cur_len = std::distance( cur->begin(), cur->end() );
+      if( prev_len > cur_len ) {
+        prev = *cur;
+        allowed_path->push_back( std::move( *cur ) );
+        prev_len = std::distance( prev.begin(), prev.end() );
+      }
+      else if( !std::equal(
+        prev.begin(), prev.end(),
+        cur->begin(), std::next( cur->begin(), prev_len )
+      ) ) {
+        prev = *cur;
+        allowed_path->push_back( std::move( *cur ) );
+        prev_len = std::distance( prev.begin(), prev.end() );
+      }
```

Comment:
I can see what you are doing here, but it definitely should be commented and explained.
Moreover, L45-47 and L53-55 are totally the same as procedures. I think these blocks need to be combined.

Reply:
fixed